### ISSUE TITLE
Add native database adapter for WatermelonDB

### ISSUE THREAD

**Issue by nchapman** [nchapman/pulsar#4]

**Comment by ospfranco** [comment]
Took a look into the sql plugin, seems bare bones enough that I can write a easy js wrapper for watermelon. I could also write some bindings myself, but maybe there is some complexity there I'm not aware of. I will just give it a go and see how far I get for a few hours.

**Comment by ospfranco** [comment]
Just took a look at the plugin and it is super weird. Instead of sticking to WebSQL or something similar, it has two functions. One is `executes`, which just executes the passed query but does not return data, and `query` which does return data. 

![Image](https://github.com/nchapman/pulsar/assets/1634213/46cd8508-a0b5-4736-8c45-2e5e2120e394)

I will try to implement the adapter with pure select queries but wdb might be expecting some actual response on some of the queries.

**Comment by ospfranco** [comment]
It seems watermelon does not support async calls. I've reached out to the author of the library to confirm this. If it really does not support async queries then it will not be possible to create a native Tauri adapter since the communication via JSON is always async due to message passing.

**Comment by ospfranco** [comment]
I'm thinking a bit why would you want to use SQLite directly. The reason for lokiJS is that browsers don't have sqlite natively, but given that Tauri uses JSON serialization, there will be little performance gain by binding to any sort of "native" API (which is what Watermelon DB does via JSI bindings, it binds JS directly to C++).

**Comment by nchapman** [comment]
@ospfranco Yeah, there are definitely some tradeoffs. lokijs is its own set of tradeoffs. I think it gives us more flexibility within the platform to have easy access to the sqlite database outside of the webview. Also I'm hopeful that over time there will be a more efficient mechanism for serialization.

Let's continue forward this as long as there aren't other blockers. Did you figure out the issue with async calls?

**Comment by ospfranco** [comment]
ok, I will get it working.

Radek is no longer active on Twitter, so he hasn't replied. The internal arch of Watermelon DB is funny and the documentation is also a bit scarce. I will continue trying to create an adapter.

**Comment by ospfranco** [comment]
Just an important FYI, the project doesn't seem to have been updated in 5 months, as well as the author Radex seems to be very unactive.

**Comment by ospfranco** [comment]
Ok, I figured out how the adapter is supposed to work. It's just really complex as there are many layers of indirection. Rough estimation would be 2 - 3 more days of work.

**Comment by ospfranco** [comment]
I'm a little overwhelmed by the complexity of WDB. 

Even though the documentation says it should only be a matter of creating a dispatcher (which in theory it is) none of the classes takes a dispatcher factory as an injectable dependency. I tried copying and pasting the adapter code inside of the project but gave up after a couple of hours, I'm ending up pulling half of the repo's code since it is all interconnected with indirection layers and abstractions and types (which are in flow and not TS). 

Tried cloning and modifying the sqlite adapter in-situ code to just use the tauri-sql plugin calls but faced many cryptic compilation errors and some vite errors because it seems it doesn't support file links/ts paths.

The complexity here is probably the reason nobody has contributed adapters for more platforms. At least to me, it seems Radek is the only one who would be able to navigate the sea of types and callbacks it takes to perform an actual sql query.

I can continue trying to get an adapter to work, but it is hard to say how much time it will take.

**Comment by ospfranco** [comment]
I've hit an error I cannot get around. I've created an [issue](https://github.com/Nozbe/WatermelonDB/issues/1757) on the WatermelonDB in the hopes someone familiar with the codebase has any idea.

**Comment by nchapman** [comment]
Thanks for pushing forward here @ospfranco! I agree that it would be much better to have support from radex to make this process as pain free as possible. He used to be quite responsive on the GitHub repo. We could make our implementation a stand alone Tauri plugin if that would incentivize them to help us out. I'm totally fine with us opening an issue there to see if we can get some guidance and putting this on hold until we can. Let me know what you think is best.
